biojava-structure-gui/src/main/java/org/biojava/nbio/structure/align/gui/jmol/JmolPanel.java

```diff
@@ -114,11 +127,28 @@ public void executeCmd(String rasmolScript) {
 		viewer.evalString(rasmolScript);
 	}
 
-	public void setStructure(Structure s)
+	public void setStructure(final Structure s)
 	{
 		this.structure = s;
-		String serialized = s.toMMCIF();
-		viewer.openStringInline(serialized);
+		try (
+				PipedOutputStream out = new PipedOutputStream();
+				// Viewer requires a BufferedInputStream for reflection
+				InputStream in = new BufferedInputStream(new PipedInputStream(out));
+				) {
+			new Thread((Runnable)() -> {
+				try {
+					MmtfActions.writeToOutputStream(s,out);
+				} catch (Exception e) {
+					logger.error("Error generating MMTF output for {}",
+							s.getStructureIdentifier()==null ? s.getStructureIdentifier().getIdentifier() : s.getName(), e);
+				}
+			}).start();
+			viewer.openReader(null, in);
+		} catch (IOException e) {
```

Comment:
Thanks for the explanation, that makes sense now.

> I guess ultimately I caught the exceptions here because they seemed very unlikely to result from user input (assuming the MMTF encoder doesn't have bugs).

Well in theory if the structure is "malformed" in some way the encoder should throw an exception saying it doesn't like it. In real life I'm not sure if such a "malformation" is a possibility.